File: packages/python/src/alumnium/clients/typecasting.py
from string import whitespace
from typing import Optional, TypeAlias, Union

Data: TypeAlias = Optional[Union[str, int, float, bool, list[Union[str, int, float, bool]]]]


def loosely_typecast(value: str | list[str]) -> Data:
    """Convert string or list of strings to appropriate Python types."""
    if isinstance(value, list):
        return [loosely_typecast(item) for item in value]  # ty:ignore[invalid-return-type]

    value = value.strip()

    if value.upper() == "NOOP":
        return None
    elif value.isdigit():
        return int(value)
    elif value.replace(".", "", 1).isdigit():
        return float(value)
    elif value.lower() == "true":
        return True
    elif value.lower() == "false":
        return False
    else:
        return value.strip(f"{whitespace}'\"")

File: packages/typescript/src/clients/typecasting.ts
export type Data =
  | null
  | string
  | number
  | boolean
  | (string | number | boolean)[];

/**
 * Convert string or list of strings to appropriate TypeScript types.
 */
export function looselyTypecast(value: string | string[]): Data {
  if (Array.isArray(value)) {
    return value.map(
      (item) => looselyTypecast(item) as string | number | boolean,
    );
  }

  const trimmed = value.trim();

  if (trimmed.toUpperCase() === "NOOP") {
    return null;
  } else if (/^\d+$/.test(trimmed)) {
    return parseInt(trimmed, 10);
  } else if (/^\d+\.\d+$/.test(trimmed)) {
    return parseFloat(trimmed);
  } else if (trimmed.toLowerCase() === "true") {
    return true;
  } else if (trimmed.toLowerCase() === "false") {
    return false;
  } else {
    // Strip whitespace and quotes
    return trimmed.replace(/^[\s'"]+|[\s'"]+$/g, "");
  }
}
